rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,574 @@
+//  - hardware key agent gRPC service implementation
+type hardwareKeyAgentService interface {
+  GetInfo(ref hardwareKeyRef) hardwareKeyInfo
+  Sign(ref hardwareKeyRef, digest []byte, opts SignerOpts) (signature []byte, err error)
+}
+
+// Implements [crypto.Signer] and [keys.HardwareSigner].
+type hardwareKeyAgentKey struct {
+  agent  keyAgentService
+  keyRef hardwareKeyRef
+  // hardwareKeyInfo is cached after the first call
+  hardwareKeyInfo hardwareKeyInfo
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Public() crypto.PublicKey {
+  return s.hardwareKeyInfo.publicKey
+}
+
+// Implement [crypto.Signer]
+func (s *agentSigner) Sign(_ io.Reader, digest []byte, opts crypto.SignerOpts) ([]byte, error) {
+  return s.agent.Sign(as.KeyRef, digest, opts)
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetAttestationStatement() *AttestationStatement {
+  return s.hardwareKeyInfo.attestationStatement
+}
+
+// Implement [keys.HardwareSigner]
+func (s *agentSigner) GetPrivateKeyPolicy() PrivateKeyPolicy {
+  return s.hardwareKeyInfo.privateKeyPolicy
+}
+
+func (s *agentSigner) fetchHardwareKeyInfo() (hardwareKeyInfo, error) {
+  if s.hardwareKeyInfo == nil {
+    s.hardwareKeyInfo = s.agent.GetInfo(s.ref)
+  }
+  return s.hardwareKeyInfo
+}
+```
+
+### UX
+
+#### PIN caching
+
+The PIN caching portion of this RFD does not present any additional UX concerns.
+It is purely a UX benefitting change, removing the need for back-to-back PIN
+prompts.
+
+#### Agent clients
+
+##### Teleport Connect
+
+Teleport Connect is an ideal runner for the key agent because it:
+
+1. Is a long-lived process
+1. Provides a UI for touch/PIN prompts, and can foreground itself for these prompts
+
+By default, if Teleport Connect detects a hardware key plugged in, it will automatically
+serve the hardware key agent service. This way, the agent will be served
+regardless of Teleport Connect's [login state](#running-the-agent-before-login).
+
+For example, if a user primarily wants to use `tsh`, but get PIN caching and
+PIN prompts in Teleport Connect, they could just launch Teleport Connect without
+logging in. Teleport Connect would just foreground itself with hardware key
+prompts for the user as needed without adding additional overhead.
+
+Alternative: By default, set `teleport.agent=false`. Once the user logs into
+Connect with a hardware key requirement for the first time, flip the flag to
+true indefinitely. The benefit with this approach is that we won't run an
+unused agent by default for users not using hardware key support. On the other
+hand, it requires the user to log in at least once for the feature to work as
+described in the example above.
+
+If desired, the agent can be disabled manually with a config option:
+
+| Property | Default | Description |
+|----------|---------|-------------|
+| `teleport.hardware_key.agent` | `false` | Starts the hardware key agent automatically |
```

Comment:
Not a big issue, but looking at [other config options we currently have](https://goteleport.com/docs/connect-your-client/teleport-connect/#configuration) (I know, the formatting in the table is horrible, I'll fix it one day), I think we're better off naming this `hardwareKeyAgent.enabled`.